include/divergence_block_chain.h

```diff
@@ -0,0 +1,166 @@
+#pragma once
+
+#include "communicator.h"
+#include "recorders.h"
+#include <mutex>
+#include <thread>
+#include <vector>
+
+namespace celerity::detail {
+/**
+ *  @brief This class is a wrapper around a 1D vector that allows us to access it as a 2D array.
+ *
+ *  It is used to send the task hashes to other nodes using MPI while keeping the code simple and readable.
+ */
+template <typename T>
+struct vector_2d {
+  public:
+	vector_2d(size_t width, size_t height) : m_data(width * height), m_width(width){};
+
+	const T& operator[](id<2> ij) const {
+		assert(ij[0] * m_width + ij[1] < m_data.size());
+		return m_data[ij[0] * m_width + ij[1]];
+	}
+
+	T* data() { return m_data.data(); }
+
+  private:
+	std::vector<T> m_data;
+	const size_t m_width;
+};
+
+/**
+ *  @brief This class gives a view into a const vector.
+ *
+ *	It is used to give us the currently unhashed task records while keeping track of the offset and width.
+ */
+template <typename T>
+struct sliding_window {
+  public:
+	sliding_window(const std::vector<T>& value) : m_value(value) {}
+
+	const T& operator[](size_t i) const {
+		assert(i >= 0 && i < m_width);
```

Comment:
```suggestion
		assert(i < m_width);
```
Can be simplified.

Also this assertion fails if `size()` is not called before, which seems weird?!